btree/btree.c

```diff
@@ -3460,21 +3535,233 @@ static int64_t btree_get_tick(struct m0_sm_op *smop)
 	};
 }
 
+/** Iterator state machine. */
 int64_t btree_iter_tick(struct m0_sm_op *smop)
 {
-	struct m0_btree_op 	*bop = M0_AMB(bop, smop, bo_op);
-	//ToDo: Implement complete destroy tick function.
-	switch(bop->bo_op.o_sm.sm_state)
-	{
-		case P_INIT:
-			return P_ACT;
+	struct m0_btree_op    *bop   = M0_AMB(bop, smop, bo_op);
+	struct td             *tree  = bop->bo_arbor->t_desc;
+	struct m0_btree_oimpl *oi    = bop->bo_i;
+	struct level          *lev;
 
-		case P_ACT:
-			return P_DONE;
+	switch (bop->bo_op.o_sm.sm_state) {
+	case P_INIT:
+		if ((bop->bo_flags & BOF_COOKIE) &&
+		    cookie_is_set(&bop->bo_rec.r_key.k_cookie))
+			return P_COOKIE;
+		else
+			return P_SETUP;
+	case P_COOKIE:
+		if (cookie_is_valid(tree, &bop->bo_rec.r_key.k_cookie))
+			return P_LOCK;
+		else
+			return P_SETUP;
+	case P_SETUP:
+		bop->bo_arbor->t_height = tree->t_height;
+		oi = level_alloc(tree->t_height);
+		if (oi == NULL)
+			return fail(bop, M0_ERR(-ENOMEM));
+		return P_LOCKALL;
+	case P_LOCKALL:
+		if (bop->bo_flags & BOF_LOCKALL)
+			return m0_sm_op_sub(&bop->bo_op, P_LOCK, P_DOWN);
+		/** Fall through if LOCKALL flag is not set. */
+	case P_DOWN:
+		oi->i_used  = 0;
+		oi->i_pivot = -1;
+		return node_get(&oi->i_nop, tree, &tree->t_root->n_addr,
+				P_NEXTDOWN);
+	case P_NEXTDOWN:
+		if (oi->i_nop.no_op.o_sm.sm_rc == 0) {
+			struct slot    s = {};
+			struct segaddr child;
 
-		default:
-			return 0;
+			lev = &oi->i_level[oi->i_used];
+			lev->l_node = oi->i_nop.no_node;
+			s.s_node = oi->i_nop.no_node;
+			lev->l_seq = lev->l_node->n_seq;
+			oi->i_key_found = node_find(&s, &bop->bo_rec.r_key);
+			lev->l_idx = s.s_idx;
+
+			if (node_level(s.s_node) > 0) {
+				if (oi->i_key_found) {
+					s.s_idx++;
+					lev->l_idx++;
+				}
+				/**
+				 * If sibling flag is set, check if the node
+				 * has valid left or right index based on next
+				 * previous flag. If valid left/right index
+				 * found. mark this level as pivot level.
+				 */
+				if (((bop->bo_flags & BOF_NEXT) &&
+				    (lev->l_idx < node_count(lev->l_node))) ||
+				    ((bop->bo_flags & BOF_PREV) &&
+				    (lev->l_idx > 0)))
+					oi->i_pivot = oi->i_used;
```

Comment:
The idea here is to update i_pivot as we traverse down the tree. The i_pivot will be updated with the i_used if this level has a valid sibling (depending on the PREV and NEXT flag) . So once we reach the leaf node, the i_pivot will be updated with level which is closes to leaf node and has valid sibling. 
This will also help in identifying if this is the rightmost key in the btree, in this case i_pivot will not get updated and reamin at -1.